src/userobjects/MoabSkinner.C

```diff
@@ -0,0 +1,1008 @@
+  }
+
+  // get variable numbers
+  _temperature_var_num = getAuxiliaryVariableNumber(_temperature_name, "temperature");
+  if (_bin_by_density)
+  {
+    if (_temperature_name == _density_name)
+      mooseError("The 'temperature' and 'density' variables cannot be the same!");
+
+    _density_var_num = getAuxiliaryVariableNumber(_density_name, "density");
+  }
+
+  if (_temperature_max <= _temperature_min)
+    paramError("temperature_max", "'temperature_max' must be greater than 'temperature_min'");
+
+  for (unsigned int i = 0; i < _n_temperature_bins + 1; ++i)
+    _temperature_bin_bounds.push_back(_temperature_min + i * _temperature_bin_width);
+
+  for (unsigned int i = 0; i < _n_density_bins + 1; ++i)
+    _density_bin_bounds.push_back(_density_min + i * _density_bin_width);
+
+  _tet4_nodes.push_back({0, 1, 2, 3});
+
+  _tet10_nodes.push_back({0, 4, 6, 7});
+  _tet10_nodes.push_back({1, 5, 4, 8});
+  _tet10_nodes.push_back({2, 6, 5, 9});
+  _tet10_nodes.push_back({7, 8, 9, 3});
+  _tet10_nodes.push_back({4, 9, 7, 8});
+  _tet10_nodes.push_back({4, 5, 9, 8});
+  _tet10_nodes.push_back({4, 7, 9, 6});
+  _tet10_nodes.push_back({4, 9, 5, 6});
+}
+
+unsigned int
+MoabSkinner::getAuxiliaryVariableNumber(const std::string & name,
+                                           const std::string & param_name) const
+{
+  if (!_fe_problem.getAuxiliarySystem().hasVariable(name))
+    paramError(param_name, "Cannot find auxiliary variable '", name, "'!");
+
+  return _fe_problem.getAuxiliarySystem().getFieldVariable<Real>(0, name).number();
+}
+
+MeshBase &
+MoabSkinner::mesh()
+{
+  return _fe_problem.mesh().getMesh();
+}
+
+void
+MoabSkinner::initialize()
+{
+  // Set spatial dimension in MOAB
+  _moab->set_dimension(mesh().spatial_dimension());
+
+  // Create a meshset representing all of the MOAB tets
+  _moab->create_meshset(moab::MESHSET_SET, _meshset);
+
+  createTags();
+
+  createMOABElems();
+
+  findBlocks();
+}
+
+void
+MoabSkinner::execute()
+{
+  if (_standalone)
+    update();
+}
+
+void
+MoabSkinner::update()
+{
+  // Clear MOAB mesh data from last timestep
+  reset();
+
+  _serialized_solution->init(_fe_problem.getAuxiliarySystem().sys().n_dofs(), false, SERIAL);
+  _fe_problem.getAuxiliarySystem().solution().localize(*_serialized_solution);
+
+  // Re-initialise the mesh data if the libMesh mesh has changed
+  if (!_fixed_mesh)
+    initialize();
+
+  // Sort libMesh elements into bins
+  sortElemsByResults();
+
+  // Find the surfaces of local temperature regions
+  findSurfaces();
+}
+
+void
+MoabSkinner::findBlocks()
+{
+  _blocks.clear();
+
+  int i = 0;
+  for (const auto & b : _fe_problem.mesh().meshSubdomains())
+    _blocks[b] = i++;
+
+  _n_block_bins = _blocks.size();
+
+  if (_material_names.size() != _n_block_bins)
+    paramError("material_names", "This parameter must be the same length as the number of "
+      "subdomains in the mesh (" + Moose::stringify(_n_block_bins) + ")");
+}
+
+void
+MoabSkinner::createMOABElems()
```

Comment:
You mean wrap this in an openmp loop?

Reply:
Meaning it might be more efficient to collect all of the data and use `_moab->create_vertices` instead of `_moab->create_vertex`. But there may be tradeoffs with that. Not something I'd worry too much about right now.